### DISCUSSION TITLE
Using `query` together with `counter.at`

### DISCUSSION THREAD

**Discussion by patrick-kidger** [typst/typst#1109] [General]
The [docs for `counter.at`](https://typst.app/docs/reference/meta/counter/#methods--at) mention using `query` to get a location. (In my case I want to get the number of pages at a certain later point in my document.)

However `query` itself returns a `content` object, not a `location` object.

How should these be combined together?

**Comment by PgBiel** [comment]
You're probably looking for `locate(loc => /* your logic using location here */)` - `locate` is a function that returns (places) a location-aware element which always provides its current location to your code (that is, provides its location as the first parameter to the callback function you give it). In other words, `loc` there will always be the location equivalent to "right here" in the document (wherever that `locate(...)` is placed). Note, however, that you will always need a `locate` to obtain a `location` in the first place - unfortunately, there is no `get-location()` function to give you a location arbitrarily (without having to wrap your code in callbacks), as Typst needs to separate location-aware from non-location-aware content due to how things work internally, so `locate(loc => ...)` will always have to be present at your top-level functions (those which return content), and then you will have to pass down the location object to pure logic functions (which don't return content), as `locate` will always return content.

However, given an existing `location` object, you can use `query` to obtain _another_ location. More specifically, you can use `query` to obtain _not_ the current location, but the location of _other elements_ in the document, given the current one. This is because `query` lets you select elements based on their types (or other selectors), and then returns all matching elements, but those elements will have a special `.location()` method (available only when returned by `query`) which gives you where that element is actually located (position and page). The catch is that `query` requires a location object (any such object) as its second parameter (the first parameter is the selector). So, you _will_ need to wrap your stuff in `locate` to use `query` either way (you can't use `query` to "escape" the requirement of using `locate` to get `location` objects).

For example, `let my-query = query(heading, loc)` will let you obtain all headings in the document. (Using `heading.before(loc)`, for example, would return you all headings _before_ that location, not in the whole document.) To get `loc` in the first place, your code will have to be inside a `locate(loc => { ...; let my-query = ...; ... })`.
Thus, `my-query` will be an array of headings, with a special property: they will have a `.location()` method. So you will be able to do, for example, `let first-heading-loc = my-query.at(0).location()`, then use the first heading's location at will. But remember: You still need _some_ location to be able to use `query`, which necessarily implies wrapping your logic within a `locate` call.

**Reply by patrick-kidger** [reply]
Hmm, I'm not sure I quite understand.

To explain what I'm trying to do. Right now I have a `#let total_pages = locate(loc=>counter(page).final(loc).at(0))` line at the top of my document.

I'd like to switch this out to get the number of pages at a particular point partway through my document. (In particular, I do not want to count the appendix.)

Do you know how this might be accomplished?

**Reply by PgBiel** [reply]
So, precisely - `#let total_pages = locate(loc=>counter(page).final(loc).at(0))` won't do what you're expecting! `#locate` is an element (much like `#heading`, `#enum`, `#block`, `#box`, `#figure`...) - `locate` does **not** return what your callback returns, but rather some content, and that content will display whichever content your callback returns. So your callback cannot return a number (or it will be converted to content and displayed literally)!  `total_pages` will _not_ be equal to a number, but rather to the `locate` element you created. `total_pages + 1` won't be a very meaningful operation, for example!

What you're looking to do is, instead of doing:

```rust
#let total_pages = locate(loc => counter(page).final(loc).at(0))

This is the total amount of pages: #total_pages
```

You want to do:

```rust
locate(loc => [
   #let total_pages = counter(page).final(loc).at(0)
   This is the total amount of pages: #total_pages
])
```

_Any_ particular code or content using a certain `loc` object **must** be called or invoked inside the `locate` that generated it, is what I'm trying to say in my comment above.

(Edit: your approach would work only to _display_ the total page amount, but `total_pages` will be a _content_ rather than a number, so you won't be able to use it for numeric operations, should you want to do that.)

 
 Regarding not counting the appendix: you can use a `query` (as you propose) to get the location of the first appendix heading, which will give you its page. Here's a sample way to do this:

```rust
locate(loc => [
   #let appendix-heading = query(heading, loc).find(h => h.body.text.starts-with("Appendix"))  // sample code
   #let appendix-loc = appendix-heading.location()
   #let appendix-page = appendix-loc.page()
   The appendix starts at Page #appendix-page.
])
```

You could complement that to get the page of the last heading before the appendix by using `#let last-heading = query(heading.before(appendix-loc), loc).last()`, and then `#let last-heading-page = last-heading.location().page()`, for example. This should give you a strong base to built on top of for your case.

**Reply by patrick-kidger** [reply]
Aha, gotcha. I only need the page count as content, so I'm doing:
```typst
#let total_pages = locate(loc=>counter(page).at(query(<before_bibliography>, loc).at(0).location()).at(0))
```
Along with adding a `<before_bibliography>` label. This works - thank you for your help.

Incidentally, it seems to me that `<before_bibliography>` behaves inconsistently depending on whether it's in a code context. (Outside of a code context: creates a label. Inside of a code context: refers to a label, much like `@before_bibliography` does outside of code mode.)

**Reply by PgBiel** [reply]
> This works - thank you for your help.

That's awesome! You're welcome!

> Incidentally, it seems to me that `<before_bibliography>` behaves inconsistently depending on whether it's in a code context. (Outside of a code context: creates a label. Inside of a code context: refers to a label, much like `@before_bibliography` does outside of code mode.)

It (that particular syntax, at least) actually behaves exactly the same - `<label-name>` creates and returns a `label("label-name")` element. The thing is that, when such element is placed in the document, the Typst interpreter will associate the placed label with the previously placed element. However, when used in code, it is not parsed as content (at least at first), so you can manipulate the label element itself. In particular, you can give it to `query`, which will use that label as a reference to find the element it looks for (it will look for the element attached to a label with the same name).

**Reply by jarkinber** [reply]
Hi @PgBiel - wanting to ask a related question here, however if you think it would be better for me to open a new issue, please let me know.

I'm trying to implement a custom outline which displays the range of pages the section covers for level 1 headings only, per the below mockup:

```
Level 1 Heading #1 ------------------------------------ 1 - 10
    Subheading 1 ------------------------------------------ 1
    Subheading 2 ------------------------------------------ 4
Level 1 Heading #2 ------------------------------------ 10 - 25
    Subheading 1 ------------------------------------------ 10
    Subheading 2 ------------------------------------------ 14
...
etc.
```

However I'm struggling to work out how to query the final page number in any one of the sections to be able to make this work:

```typst
show outline.entry.where(
    level: 1
  ): it => {
    set text(size: 11pt, weight: "bold")
    if it.at("label", default: none) == <modified-entry> {
      it // prevent infinite recursion
    } else {
      [
        #outline.entry(
          it.level,
          it.element,
          it.body,
          it.fill,
          locate(loc => {
            let total_pages = counter(page).final(loc).at(0) 
            // ^^Looking for something like what you suggest for the OP, but instead of the final page in the document, the final page in that specific section?
            it.page + " - " + [#total_pages]
          }
          )
        ) <modified-entry>
      ]
    }
  }
```

Do you have any suggestions for how I might achieve this?

Thanks for your help!

**Reply by PgBiel** [reply]
Hey @jarkinber , I made the following code which will work on 0.11.0+ (not yet released, but you can test it in the web app at the moment):

```typ
#let last-section-page-state(n) = state("last-section-page-" + str(n), 0)

#show outline.entry.where(level: 1): it => {
  let num = counter(heading).at(it.element.location()).first()
  [#it.body#box(width: 1fr, it.fill)#it.page .. #context last-section-page-state(num).final()]
}

#outline()

#set heading(numbering: "1.")
#show heading.where(level: 1): it => {
  if it.numbering != none {
    let num = counter(heading).get().first()
    if num > 1 {
      // Save where the previous section ended
      last-section-page-state(num - 1).update(here().page())
    }
  }

  it
}

#context {
  // Ensure the last section ends at the last page
  last-section-page-state(counter(heading).final().first()).update(counter(page).final().first())
}

// ---

= Hello world!

== Ok thanks

#pagebreak()

== Ok thanks

= Bye world!

== Bye thanks

== Bye thanks

#pagebreak()

asdfadsfsdf

#pagebreak()

asdfasdfasdfa

= Cool

aaa
#pagebreak()
```

![output](https://github.com/typst/typst/assets/9021226/35cbe848-8d75-4519-84da-e67272b03a4b)

For compatibility, I present a version for Typst 0.10.0 and below as well. (I just had to do it in 0.11.0+ syntax because the new `context` feature, replacing `locate` and the likes, is very handy.)

<details><summary>Older version (using <code>locate</code> instead of <code>context</code>)</summary>

<p>

```typ
#let last-section-page-state(n) = state("last-section-page-" + str(n), 0)

#show outline.entry.where(level: 1): it => {
  let num = counter(heading).at(it.element.location()).first()
  [#it.body#box(width: 1fr, it.fill)#it.page .. #locate(loc => last-section-page-state(num).final(loc))]
}

#outline()

#set heading(numbering: "1.")
#show heading.where(level: 1): it => {
  if it.numbering != none {
    let num = counter(heading).at(it.location()).first()
    if num > 1 {
      // Save where the previous section ended
      last-section-page-state(num - 1).update(it.location().page())
    }
  }

  it
}

#locate(loc => {
  // Ensure the last section ends at the last page
  last-section-page-state(counter(heading).final(loc).first()).update(counter(page).final(loc).first())
})

// ---

= Hello world!

== Ok thanks

#pagebreak()

== Ok thanks

= Bye world!

== Bye thanks

== Bye thanks

#pagebreak()

asdfadsfsdf

#pagebreak()

asdfasdfasdfa

= Cool

aaa
#pagebreak()
```

</p>

</details>

**Reply by jarkinber** [reply]
Hi @PgBiel, thank you so much for the swift and detailed response! Very much appreciated. Confirming that I can replicate that functionality both in the web app as well as using the legacy method. Do you think this kind of functionality is something that could be worth implementing in a more concrete way as an outline customisation?

**Reply by PgBiel** [reply]
If you mean the possibility of having this built-in, I'm not really sure. Could perhaps be nice to have the option to do it, but it could also just be a better idea to delegate this job to a package, depending on how much effort would be needed to implement and maintain such a feature in the compiler. Either way, you can open an issue to gather further opinions.
